@eessex => [Error Handling] Only serve non 2XX pages on principalField errors

mzikherman opened the pull request:
aka unleash partial renders.

Doing some testing of this locally, and it seems to do what we want!

I ran Reaction linked with Force, pointing to local Metaphysics.

I then went into the resolvers for some of the leafs used in the artwork page, and intentionally threw errors there (to mimic an upstream service/fetch failing).

Intentionally threw an error when fetching some of the grids in 'Other Works', threw an error when fetching sale data and threw an error when fetching Vortex price histogram data.

In all cases without this change, a 500 was served. With this change, the page rendered with a 200, and at least in these specific examples, there didn't seem to be any negative affect on UI/layout.

Some info was just missing, an entire artwork grid in 'Other Works' was gone, but the page seemed pretty happy to me! There were no null/exist-y related crashes.

I let this as a [HOLD] b/c this unleashes partial renders.

mzikherman commented on src/Artsy/Router/Utils/Route.tsx:
```diff
@@ -28,33 +27,8 @@ function createRender({
-      // second argument.
-      // @ts-ignore
-      throw new HttpError(status, message)
+    if (error && error instanceof HttpError) {
```
We throw an error of this type [here](https://github.com/artsy/reaction/blob/424e58addba47af4ed30ecedc753d32f100d4049/src/Artsy/Relay/createRelaySSREnvironment.ts#L115), only when there is an underlying principal field error w/ a valid HTTP status code.

Otherwise, no more throwing!

damassi commented on src/Artsy/Router/Utils/Route.tsx:
```diff
@@ -28,33 +27,8 @@ function createRender({
-      // @ts-ignore
-      throw new HttpError(status, message)
+    if (error && error instanceof HttpError) {
+      throw error
```
🎉

damassi replied:
Note that we need 
```
throw new HttpError(status, message)
```
For errors to be redirected to the error page within the router: https://github.com/4Catalyzer/found#error-handling

mzikherman replied:
This `error` is already an `HttpError` with the appropriate status, from https://github.com/artsy/reaction/blob/c7d112dc3fc53da07c191ebbc1cf6e7fb01fd48d/src/Artsy/Relay/createRelaySSREnvironment.ts#L115 (`httpError` is probably not a great variable name, that is the status code. We don't pass in the `message`).

So this all should work (ie-  a principal field 404 by going to a non existent artist/artwork page correctly returns a 404 and renders that error page).

eessex commented:
Which tickets are blockers to releasing this?

mzikherman commented:
The only ticket is really https://artsyproduct.atlassian.net/secure/RapidBoard.jspa?rapidView=64&projectKey=PLATFORM&modal=detail&selectedIssue=PLATFORM-1740 .

Basically, with a lot more partial renders happening, we'll have more potential for components to error and so we should set ourselves up for success by making sure Sentry is tidy and we're satisfied with how errors are [currently] reported.

Additionally, since the overall Force 5XX rate will be affected, we should make sure we're satisfied with how that is measured. I suppose that can be another card, although that should already be pretty easily available to check out before/after deploys, on staging, etc.

There isn't anything that _technically_ needs to happen in order to launch this, it's more just about feeling confident.

alloy commented:
This probably warrants a major version bump, as it _might_ lead to more runtime crashes. (And honestly, is also a milestone, imo.)

damassi commented on src/Artsy/Relay/createRelaySSREnvironment.ts:
```diff
@@ -76,63 +76,69 @@ export function createRelaySSREnvironment(config: Config = {}) {
+        loggingEnabled && metaphysicsExtensionsLoggerMiddleware(),
+        loggingEnabled && errorMiddleware({ disableServerMiddlewareTip: true }),
+      ],
+      { noThrow: true }
```
We'll want a big bold `NOTE` / `TODO` pointing at the comment in the readme about 
> noThrow - EXPERIMENTAL (May be deprecated in the future) set true to not throw when an error response is given by the server, and to instead handle errors in your app code.

mzikherman replied:
Yea, I will clean this PR up and pop into the Relay slack to see about this.

damassi commented on src/Artsy/Relay/createRelaySSREnvironment.ts:
```diff
@@ -76,63 +76,69 @@ export function createRelaySSREnvironment(config: Config = {}) {
+          } else {
+            return res
+          }
+        },
```
Can this be moved into a function in the [middleware](https://github.com/artsy/reaction/tree/master/src/Artsy/Relay/middleware) folder? 

(And if there's any other custom stuff in here that can be moved, that would be good too.)

damassi commented on src/Artsy/Relay/createRelaySSREnvironment.ts:
```diff
@@ -76,63 +76,69 @@ export function createRelaySSREnvironment(config: Config = {}) {
+          } else {
+            const error = new NetworkError(response.statusText)
+            error.response = response
+            throw error
```
Like this -- something like `errorHandlerMiddleware.ts` (or whatever)

damassi commented on src/Artsy/Relay/middleware/searchBarImmediateResolveMiddleware.ts:
```diff
@@ -0,0 +1,10 @@
+      return Promise.resolve({ data: { viewer: {} } })
+    return next(req)
+  }
+}
```
Thanks for moving these into functions; this network file is much cleaner now 👍

mzikherman commented:
Addressed the feedback, this might be mergeable!

I double-checked that 500s on Force locally do show up in staging, so I think we're covered in terms of monitoring. Unleash the partial renders!

We can also merge this and let it sit in staging for a little bit.

mzikherman commented:
Ok, going for it, based on our technical plan all the steps/prep work are done!

artsyit commented:
<!-- GITHUB_RELEASE COMMENT: released -->
:rocket: PR was released in v20.0.0 :rocket:
